### PR TITLE
Fix audio crash

### LINKED ISSUE #79937: Audio update quit to desktop after moving 1-2 tiles.
### Describe the bug

probable cause #79812 

Not sure what is going on honestly, all i know is that playing with this update my game just quits the moment i start walking about, it doesn't if it stand still though.

### Attach save file

[Zakhadia_BigCity-trimmed.tar.gz](https://github.com/user-attachments/files/19042674/Zakhadia_BigCity-trimmed.tar.gz)

### Steps to reproduce

Load save with a version with bdf114b, or build 2025-03-01-1406, game works fine.

Update to build 2025-03-01-1634 or later, the game will quit to desktop after moving the character a couple of tiles. 

### Expected behavior

Not to quit to desktop?

### Screenshots

_No response_

### Versions and configuration

- OS: Windows
    - OS Version: 7
- Game Version: bdf114b [64-bit]
- Graphics Version: Tiles
- Game Language: System language []
- Mods loaded: [
    Dark Days Ahead [dda],
    Stats Through Skills [StatsThroughSkills],
    SpeedyDex [speedydex],
    Disable NPC Needs [no_npc_food],
    Medieval Mod Reborn [Medieval_Mod_Reborn],
    Tamable Wildlife [Tamable_Wildlife],
    Mind Over Matter [mindovermatter],
    Bombastic Perks [bombastic_perks],
    Slowdown Fungal Growth [no_fungal_growth]
]

### Additional context

[debug.log](https://github.com/user-attachments/files/19042684/debug.log)

There's no crashlog because it isn't being produced when the game quits it's not crashing it's quitting, the only one i have is for a few days before.

### PR DESCRIPTION
**Pull request by redpenguiney**
#### Summary
None

#### Purpose of change

Fixes #79937 (my bad). Hopefully.

#### Describe the solution

Apparently you can't call Mix_HaltAudio() inside the audio callback, so the callback now just adds the audio channels that need to be halted to a list, which gets cleared the next time someone tries to play a sound effect.

#### Describe alternatives you've considered

Could have just made it loop 4 times as long as it needed to (since when time is slowed audio takes 4x as long to play) and not manually halted it at all, but that would cause most audio channels to be wasted playing absolutely nothing for most of the time.

#### Testing

I walked around doing various things that produce sound, some of which while time was slowed.
That said, I couldn't reproduce #79937 on my system to begin with, so other people definitely need to test this too.

#### Additional context
None

### CODE REVIEW

**Comment by redpenguiney** [comment]
FYI, I'm not sure why github made me do some weird manual merging stuff, and it's possible I messed something up while doing that.

**Comment by PatrikLundell** [comment]
The reason you're forced to do manual merging is that master has changed between the time the master you started from was generated to the time the PR is posted, and it tries to run the tests against some kind of automatic merge of PRs to the master at the time of the test. This also means that if your PR isn't merged for a long time, you hay have to perform several manual meres to sort out conflicts before it's finally merged (there are several reasons for why it can take a long time, but that's a different set of issues).

**Comment by akrieger** [comment]
Also need to remove cc-sounds.

**Comment by Wojciech79** [comment]
> can't fix the style errors and what not right now, will do so in like 8 hours

When it will be fixed?

**Comment by redpenguiney** [comment]
Alright, I've finally made the PR only modify the file it's supposed to modify (I'm starting to know how to use git now, which is nice). Also, is this the sort of situation where squashing the 30 poorly named commits would make sense?

**Comment by redpenguiney** [comment]
Why is this PR labelled with a bunch of inaccurate labels?
(also of course gcc 9 doesn't have std::fmodf inside <cmath> so it fails the very last test)

**Comment by akrieger** [comment]
> is this the sort of situation where squashing the 30 poorly named commits would make sense?

Yes.

**Comment by akrieger** [comment]
> Why is this PR labelled with a bunch of inaccurate labels?

Because the pr touched a bunch of files and labels are generally only added never removed.

File: src/sdlsound.cpp [hunk 1]
```diff
@@ -689,6 +690,14 @@ struct sound_effect_handler {
...
         0; // with respect to audio_src, in samples. for fractional indices, the output is interpolated between the two closest samples
     int loops_remaining = 0;
+    bool marked_for_termination = false;
+
+    // Because we're not allowed to call Mix_HaltChannel inside audio callbacks, slowed_time_effect() instead just adds the channel the sound is playing on to this list.
+    // whenever make_audio() is called, it will halt any channels in this list.
+    static inline std::vector < sfx::channel > channels_to_end;
+
+    // need a mutex so that make_audio() and slowed_time_effect() don't modify channels_to_end simultaneously
+    static inline std::mutex channels_to_end_mutex;
```
Comment (akrieger) [inline src/sdlsound.cpp:700]:
```suggestion
    // whenever make_audio() is called, it will halt any channels in this list.
    static std::vector < sfx::channel > channels_to_end;

    // need a mutex so that make_audio() and slowed_time_effect() don't modify channels_to_end simultaneously
    static std::mutex channels_to_end_mutex;
```

File: src/sdlsound.cpp [hunk 2]
```diff
@@ -714,63 +724,73 @@ struct sound_effect_handler {
...
 
-            for( int ear_offset = 0; ear_offset < 4;
-                 ear_offset += 2 ) { // have to handle each ear seperately for stereo audio
-                sample low_value;
-                sample high_value;
+        // check if the sound effect should end bc we done looping
+        //  (note: we let the sound loop an extra time here because when handler->loops_remaining == -1,
+        //  while that means we're done SENDING audio, that doesn't mean the audio device is done PLAYING it).
+        if( !handler->marked_for_termination && handler->loops_remaining < -1 ) {
+            if( channels_to_end_mutex.try_lock() ) { // try_lock(); we do NOT want the audio thread to block
```
Comment (akrieger) [inline src/sdlsound.cpp:732]: So what happens if we fail to lock here? We go and continue playing this effect even though it's supposed to be dead?

Comment (redpenguiney) [reply src/sdlsound.cpp:732]: notice that we write 0 to the output buffer if there are 0 loops remaining, and it'll try again. (performance impact should be negligible since the odds of the main thread being in that function/using the mutex when it needs to access it is very low anyways)

File: src/sdlsound.cpp [hunk 3]
```diff
@@ -714,63 +724,73 @@ struct sound_effect_handler {
...
+                     ear_offset += 2 ) { // have to handle each ear seperately for stereo audio
+                    sample low_value;
+                    sample high_value;
+
+                    if( handler->loops_remaining > -1 ) {
+                        memcpy( &low_value, static_cast<uint8_t *>( handler->audio_src->abuf ) + ear_offset + low_index *
+                                bytes_per_sample, sizeof( sample ) );
+                    } else {
+                        low_value = 0;
+                        std::cout << "";
```
Comment (redpenguiney) [inline src/sdlsound.cpp:768]: whoops, i better remove this when i get back to my  work station

File: src/sdlsound.cpp (hunk 1 above)
Comment (akrieger) [reply src/sdlsound.cpp:700]: Tbh I removed it because I can't recall ever seeing static inline member _variables_ before, although as a way of avoiding declaring explicit storage that makes sense, I'd rather see them as statics in namespace{} because there's zero reason they need to be members of this class.

File: src/sdlsound.cpp (hunk 2 above)
Comment (akrieger) [reply src/sdlsound.cpp:732]: > we write 0 to the output buffer if there are 0 loops remaining

Where?

File: src/sdlsound.cpp (hunk 1 above)
Comment (redpenguiney) [reply src/sdlsound.cpp:700]: I put them as static members since it makes it clear that the data only matters to sound_effect_handler. i'll move them to an anonymous namespace if you want.

File: src/sdlsound.cpp (hunk 2 above)
Comment (redpenguiney) [reply src/sdlsound.cpp:732]: i'm realizing that now, yeah. once i get back to my workstation i'll try to clean it up and document it

**Comment by PestilenceUwU** [comment]
hello, new to the community and not the most tech savvy, how long would it be until this fix is actually implemented? sorry if its rude to ask, its just a tad frustrating being unable to play on newly updated versions.

**Comment by Daved27hundred** [comment]
> hello, new to the community and not the most tech savvy, how long would it be until this fix is actually implemented? sorry if its rude to ask, its just a tad frustrating being unable to play on newly updated versions.

I'd recommend reverting your game version to before #79812 was merged until then. that'd be this build right here according to my github desktop:
image
------------------------------------------------------------------------
type in that string of numbers in the "find build" box of your launcher (I'm using kitten, it should be similar to other launchers) and hit "add to list", then "update game." You should be good to go, I'd recommend starting a new save file in a separate world so you don't risk breaking your existing save.

image

**Comment by RenechCDDA** [comment]
@PestilenceUwU Please refrain from commenting in PRs asking when other people will do things for you. It is, at best, extremely annoying.

Everyone working on this game is a volunteer. We don't get paid. We don't have schedules for when we do this. We all have lives, and nobody is forcing you to play on a specific version. If you cannot deal with that, do not play on the versions you have issues with. Our official recommendation is that you should be playing *stable*, where this sort of issue will never occur.

**Comment by PestilenceUwU** [comment]
@RenechCDDA Okay, my apologies, it wont happen again. thank you for the continued work on the game!

**Comment by RenechCDDA** [comment]
Taking a step back for a moment, since we're already annoying everyone with notifications:

The automatic tests are still running, or rather the author just asked them to be re-ran because of one of them failed. The test failure was unrelated, but we'd generally like to see all the tests pass before something like this gets merged. We don't want to break *more* stuff because we were in a hurry - that doesn't help anybody.

Past that, a merger needs to look at it and will generally build and test it in some manner to make sure it works as claimed. At the very least they need to look at what it does and press the big red button. There's a handful of mergers, and they really do have lives. Reviewing PRs, especially in our gigantic codebase, is not easy. It's not something you can do during a TV commercial.

If you want to know more about how how github pull requests generally go, [github has some details on it.](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/about-pull-requests) If you have questions beyond that you can ask in the development discord, or even raise an issue in this repository [in the issues section](https://github.com/CleverRaven/Cataclysm-DDA/issues). Those are all fine choices. 

Just please, not in PRs. PRs are "work space", right? Getting notifications for things they're working on or reviewing is generally  supposed to be a new commit, a merger approval, something like that. So if you're trying to work on making the game better, getting those comments/unrelated questions sent directly to your inbox is very distracting.

Okay! Lot of words, back to business.

**Comment by redpenguiney** [comment]
Just so it doesn't get forgotten after all these other comments and what not, someone who actually experienced the crash on their machine needs to confirm that this fix actually works; I was never able to reproduce this crash on my machine so while I fixed the (presumed) cause, I can't be sure the crash is really gone.

**Comment by KheaHyena** [comment]
> Just so it doesn't get forgotten after all these other comments and what not, someone who actually experienced the crash on their machine needs to confirm that this fix actually works; I was never able to reproduce this crash on my machine, so while I fixed the (presumed) cause, I can't be sure the crash is really gone.

@redpenguiney I know this **Pull** has been merged already but since you asked for a heads-up on confirmation, I can tip in since I had quite the hefty crashes on my current save.
The crashes appear to be gone, but it also doesn't run quite as expected. Now instead of the crash, sound just simply stops. Music is fine and continues to play, but all sound effects themselves are gone. They start out fine but I'm assuming that instead of the crash now, it simply silences all sound effects.

I will probably put up an **Issue** later on if needed, as I test a bit more.
Still, thank you for your time on putting up this fix!

**Comment by zakhad** [comment]
Limited testing feedback so far from 5 minutes of playing, game loads, game plays, seems fine, after a few minutes not sure what's causing it, but my soundeffects just stop, music carries on, then it gave me an error when i went to save/quit to see if reloading fixed it but it quit to desktop, i am able to reload my save, and soundeffects continue to work so far, so something is failing after a while.

**Comment by redpenguiney** [comment]
Aw, man! At least it's playable again in the meantime. 

Are sound effects being cut off in the middle of playing, or is it just new ones failing to play?
Also, I'm not sure how the function dbg(error) works but I'm pretty sure any SDL_Mixer calls that fail should get logged, so if one of you could send logs that would be great.

**Comment by zakhad** [comment]
[debug.log](https://github.com/user-attachments/files/19146387/debug.log)

(continued from above) ERROR : D:\a\Cataclysm-DDA\Cataclysm-DDA\src\sdlsound.cpp:983: Failed to play sound effect: No free channels available id:plmove variant:clear_obstacle season:summer
23:14:14.946 ERROR SDL : D:\a\Cataclysm-DDA\Cataclysm-DDA\src\sdlsound.cpp:983: Failed to play sound effect: No free channels available id:plmove variant:clear_obstacle season:summer
23:14:16.589 ERROR SDL : D:\a\Cataclysm-DDA\Cataclysm-DDA\src\sdlsound.cpp:983: Failed to play sound effect: No free channels available id:plmove variant:walk_tarmac season:summer
23:27:51.380 ERROR SDL : (error message will follow backtrace)
  #0
    (dbghelp: debug_write_backtrace+0xb5@000000013F922125[cataclysm-tiles.exe+0x592125]), 
  #1
    (dbghelp: DebugLog+0x235@000000013F920C65[cataclysm-tiles.exe+0x590c65]), 
  #2
    (dbghelp: sfx::play_variant_sound+0x233@000000013FDB9083[cataclysm-tiles.exe+0xa29083]), 
  #3
    (dbghelp: <lambda_6000464dbd3b232c6f4afe186bbbea28>::operator()+0x71@000000013FEB50B1[cataclysm-tiles.exe+0xb250b1]), 
  #4
    (dbghelp: sfx::do_footstep+0xcdf@000000013FEB9DEF[cataclysm-tiles.exe+0xb29def]), 
  #5
    (dbghelp: Character::make_footstep_noise+0x661@000000013FC7F251[cataclysm-tiles.exe+0x8ef251]), 
  #6
    (dbghelp: game::walk_move+0x29bd@000000013FA03D2D[cataclysm-tiles.exe+0x673d2d]), 
  #7
    (dbghelp: avatar_action::move+0x292a@00000001407B0BDA[cataclysm-tiles.exe+0x1420bda]), 
  #8
    (dbghelp: game::do_regular_action+0x763@000000013FF05453[cataclysm-tiles.exe+0xb75453]), 
  #9
    (dbghelp: game::handle_action+0xa17@000000013FF0FFE7[cataclysm-tiles.exe+0xb7ffe7]), 
  #10
    (dbghelp: do_turn+0xd89@000000013F92A839[cataclysm-tiles.exe+0x59a839]), 
  #11
    (dbghelp: WinMain+0xeb5@000000013F7A9955[cataclysm-tiles.exe+0x419955]), 
  #12
    (dbghelp: __scrt_common_main_seh+0x106@000000014100C442[cataclysm-tiles.exe+0x1c7c442]), 
  #13
    (dbghelp: BaseThreadInitThunk+0xd@000000007758556D[kernel32.dll+0x1556d]), 
  #14
    (dbghelp: RtlUserThreadStart+0x1d@00000000776E372D[ntdll.dll+0x5372d]), 
Backtrace emission took 0 seconds.
(continued from above) ERROR : D:\a\Cataclysm-DDA\Cataclysm-DDA\src\sdlsound.cpp:983: Failed to play sound effect: No free channels available id:plmove variant:walk_tarmac season:summer
23:27:58.721 ERROR SDL : D:\a\Cataclysm-DDA\Cataclysm-DDA\src\sdlsound.cpp:983: Failed to play sound effect: No free channels available id:plmove variant:clear_obstacle season:summer
23:27:59.884 ERROR SDL : D:\a\Cataclysm-DDA\Cataclysm-DDA\src\sdlsound.cpp:983: Failed to play sound effect: No free channels available id:plmove variant:clear_obstacle season:summer

is basically being repeated.

**Comment by redpenguiney** [comment]
Those errors are what I suspected. It's not properly freeing up channels when a sound effect finishes. It's strange that this would happen even if you stand still since no sound effects should be playing then. I'll make another PR with a fix soon.

**Comment by zakhad** [comment]
Did you see the weather edit i made, in that there's no weather sound effects either, despite having them in my soundpack folder, and they play fine and did previously.

**Comment by redpenguiney** [comment]
Yes, I'll take a look at weather audio too.

**Comment by Celepito** [comment]
FWIW, with build 2025-03-08-1527, I'm still randomly crashing to desktop without any error message or log, which I assume is still related to this issue, since using the build before this was added works without crashing.

This crash exists with newly created worlds with only the base standard mods added.

**Comment by redpenguiney** [comment]
> FWIW, with build 2025-03-08-1527, I'm still randomly crashing to desktop without any error message or log, which I assume is still related to this issue, since using the build before this was added works without crashing.
> 
> This crash exists with newly created worlds with only the base standard mods added.

Uh oh. That probably means the halting in the audio callback wasn't the only bad thing I was doing. Could you say what OS/compiler you're on? If I'm lucky enough to have access to a similar device it would be a godsend while debugging.

**Comment by Celepito** [comment]
The Game report spits out:

- OS: Windows
    - OS Version: 10.0.19044.3086 (21H2)

**Comment by redpenguiney** [comment]
> FWIW, with build 2025-03-08-1527, I'm still randomly crashing to desktop without any error message or log, which I assume is still related to this issue, since using the build before this was added works without crashing.
> 
> This crash exists with newly created worlds with only the base standard mods added.

OH, thank god, this PR wasn't in build 2025-03-08-1527, you need build 2025-03-08-1836. That crash _should_ be gone if you update.

That said, it's very curious that you're experiencing this on Windows 10, since that's what I was building and testing this on. Were you using the one built on MSVC or something else?

**Comment by Celepito** [comment]
Wait, what. I literally updated after that fix was added and appeared in the launchers changelog, I double checked! -.-

Lemme update and test.

**Comment by Celepito** [comment]
Okay, so I spend ~12 in game hours walking and waiting (waiting in 2h increments), I havent crashed, and can confirm the stopped audio issue (with trying to sleep bringing them back temporarily). If I CTD in the future, I'll open a new issue, but I would think that part of it should be fixed now, I hope. Sorry for the confusion, no clue why I got the wrong build.

**Comment by redpenguiney** [comment]
figured it out, when I tried to make the control flow clearer, i also made it completely incorrect. PR to fix it will be dropping in an hour or so.

**Comment by KheaHyena** [comment]
> Wait, what. I literally updated after that fix was added and appeared in the launchers changelog, I double checked! -.-
> 
> Lemme update and test.

@Celepito Sometimes the Launcher disallows the update of the most recent build for a while (or a long while). It will display the changelog, it will even say there's a new version, but if you check the dropdown box to confirm, you will see the most recent build greyed out. 
I had to wait for an hour or so until I could get the latest build to do the testing myself.

As for the testing itself, I still didn't have any crashes as before but no sound played as reported before, other than the Wind to indicate one is outside and such. Had a crash when Saving & Exiting among other similar issues (the Save wasn't lost curiously enough), but I believe that's due to other problems already reported.
I will check up on your new fix after merge.

**Comment by drauci** [comment]
I installed the latest build cdda-windows-with-graphics-and-sounds-x64-2025-03-09-0542... I have safe from of the world I created with previus build... It works for some time but after a while sound stop working music plays...

**Comment by That1Git** [comment]
> > FWIW, with build 2025-03-08-1527, I'm still randomly crashing to desktop without any error message or log, which I assume is still related to this issue, since using the build before this was added works without crashing.
> > This crash exists with newly created worlds with only the base standard mods added.
> 
> OH, thank god, this PR wasn't in build 2025-03-08-1527, you need build 2025-03-08-1836. That crash _should_ be gone if you update.
> 
> That said, it's very curious that you're experiencing this on Windows 10, since that's what I was building and testing this on. Were you using the one built on MSVC or something else?

Not sure if this breaks GitHub etiquette, but I suppose it might be helpful. I'm using Windows, standard game. Didn't build it from source. Also got crashes before updating. Is it related to hardware? I got crashes less often on my Linux machine, but that one's also more powerful.

Laptop is a Dell Latitude 3140, running Windows 11 enterprise, and 8 GB of ram. Processor is an Intel N200.

**Comment by zakhad** [comment]
@redpenguiney it's been a couple of days, haven't had any crashes, but i get no weather sound effects, or out of breath sound effects either, but portal storm weather, sound effects do work/play.

**Comment by redpenguiney** [comment]
> @redpenguiney it's been a couple of days, haven't had any crashes, but i get no weather sound effects, or out of breath sound effects either, but portal storm weather, sound effects do work/play.

That's strange, I previously tested weather specifically and had no issues. I'll test that and the out of breath sounds again. I think I can imagine an edge case where using a prespecified/reserved channel (like for weather or out of breath) could make the current system mad, so I'll check if that is actually a problem and report back.

**Comment by redpenguiney** [comment]
> @redpenguiney it's been a couple of days, haven't had any crashes, but i get no weather sound effects, or out of breath sound effects either, but portal storm weather, sound effects do work/play.

I was not able to reproduce this. Did you miss #80084, which fixed some other audio issues caused by me?

**Comment by zakhad** [comment]
@redpenguiney No i am up to date, and have been every day pretty much, it's still the case no weather no out of breath sound effects, everything else is fine.

**Comment by RenechCDDA** [comment]
@zakhad Whatever audio issue you may or may not have is not relevant to this PR and should be opened as an issue if anything.
